lib/datadog/tracing/contrib/rails/auto_instrument_railtie.rb

```diff
@@ -1,10 +1,11 @@
 # typed: ignore
+require 'datadog/tracing/contrib/auto_instrument'
 
 # Railtie to include AutoInstrumentation in rails loading
 class DatadogAutoInstrumentRailtie < Rails::Railtie
   # we want to load before config initializers so that any user supplied config
   # in config/initializers/datadog.rb will take precedence
   initializer 'datadog.start_tracer', before: :load_config_initializers do
```

Comment:
Wondering what happens if we're past `:load_config_initializers` stage:
- does the block still run?
- if not, are we able to detect that and call the block?

Reply:
This is what I tested and the answer is no: it does not run.

When `require 'datadog/tracing/contrib/auto_instrument'` is loaded, if `Contrib::Rails::Utils#railtie_supported?` evaluates to `true`, then this `Railtie` is supposed to be loaded, under the assumption that it will handle the auto-instrumentation at the appropriate time. Problem is the `Railtie` is unable to do this when auto-instrumentation is triggered in an initializer file... initialization has already begun. A "chicken and egg" problem.

We could change the logic of `Contrib::Rails::Utils#railtie_supported?` to return `false` if initialization has already begun. Or we could change `auto_instrument_railtie.rb` to skip the Railtie and `auto-instrument` directly if initialization has already begun. Either would trigger `AutoInstrument.patch_all`.

Only downside of doing so (I can think of) is in the case of:

```ruby
require 'ddtrace'
require 'ddtrace/auto_instrument`

Datadog.configure do |c|
  c.service = 'my-app'
end
```

...it will trigger a two builds of components, because instrumentation is done through `configure`. Remember how we talked about this as a negative side-effect? Well, here it is again. A reason for moving instrumentation activation out of configuration.

If we implemented a `Datadog::Tracing.instrument` API then you can have auto-instrumentation trigger that instead and skip the entire configuration API and lifecycle.

Otherwise, one potential solution is get rid of `require` and force use of a configuration option instead:

```ruby
require 'ddtrace'

Datadog.configure do |c|
  c.service = 'my-app'

  # Auto-activate all instrumentation
  c.auto_instrument

  # OR

  # Auto-activate tracing instrumentation
  c.tracing.auto_instrument
end
```